apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs

```diff
@@ -0,0 +1,393 @@
+    );
+
+    apply_derived_tx_aux(store, events, last_epoch, active_validators)
+}
+
+fn apply_derived_tx_aux(
+    store: &mut impl Store,
+    events: Vec<MultiSignedEthEvent>,
+    last_epoch: Epoch,
+    active_validators: BTreeSet<WeightedValidator<Address>>,
+) -> Result<TxResult> {
+    let total_voting_power = utils::sum_voting_powers(&active_validators);
+    tracing::debug!(
+        ?total_voting_power,
+        epoch = %last_epoch,
+        "got total voting power for epoch"
+    );
+
+    let voters = utils::get_voters_for_events(events.iter());
+    tracing::debug!(?voters, "Got validators who voted on at least one event");
+    let voting_powers =
+        utils::get_voting_powers_for_selected(&active_validators, voters)?;
+    tracing::debug!(
+        ?voting_powers,
+        "got voting powers for relevant validators"
+    );
+
+    let updates = events.into_iter().map(Into::<EthMsgUpdate>::into).collect();
+
+    let changed_keys =
+        apply_updates(store, updates, total_voting_power, voting_powers)?;
+
+    Ok(TxResult {
+        changed_keys,
+        ..Default::default()
+    })
+}
+
+/// Apply an Ethereum state update + act on any events which are confirmed
+pub(super) fn apply_updates(
+    store: &mut impl Store,
+    updates: HashSet<EthMsgUpdate>,
+    total_voting_power: VotingPower,
+    voting_powers: HashMap<Address, VotingPower>,
+) -> Result<BTreeSet<storage::Key>> {
+    tracing::debug!(
+        updates.len = updates.len(),
+        %total_voting_power,
+        ?voting_powers,
+        "Applying Ethereum state update transaction"
+    );
+
+    let mut changed_keys = BTreeSet::default();
+    let mut confirmed = vec![];
+    for update in updates {
+        // The order in which updates are applied to storage does not matter.
+        // The final storage state will be the same regardless.
+        let (mut changed, newly_confirmed) = apply_single_update(
+            store,
+            update.clone(),
+            total_voting_power,
+            &voting_powers,
+        )?;
+        changed_keys.append(&mut changed);
+        if newly_confirmed {
+            confirmed.push(update.body);
+        }
+    }
+    if confirmed.is_empty() {
+        tracing::debug!("No events were newly confirmed");
+        return Ok(changed_keys);
+    }
+    tracing::debug!(n = confirmed.len(), "Events were newly confirmed",);
+
+    // Right now, the order in which events are acted on does not matter.
+    // For `TransfersToNamada` events, they can happen in any order.
+    for event in &confirmed {
+        let mut changed = events::act_on(store, event)?;
+        changed_keys.append(&mut changed);
+    }
+    Ok(changed_keys)
+}
+
+/// Apply an [`EthMsgUpdate`] to storage. Returns any keys changed and whether
+/// the event was newly seen.
+fn apply_single_update(
+    store: &mut impl Store,
+    update: EthMsgUpdate,
+    total_voting_power: VotingPower,
+    voting_powers: &HashMap<Address, VotingPower>,
```

Comment:
We work out the fractional voting power in the function itself (using `total_voting_power` as the denominator), though not using the `FractionalVotingPower` type (yet). I'll try to make this clearer the voting powers here are absolute and should be for the same epoch i.e.

`total_voting_power` - total voting power of all the active validators for epoch `n`
`voting_powers` - the absolute voting power for all active validators for epoch `n` who submitted an Ethereum event in their vote extension

Reply:
That part is fine, but when I look at the unimplemented calculate_diff function, we don't have all the necessary info being passed in. This is because the previous voting powers may have come from a different epoch, but I only see the total_voting_power for the current epoch being passed int.

Reply:
And to get that data, it think it is easier to change the signatures all the way up and use normalized voting powers (i.e. fractional voting powers)